Firestore/core/test/firebase/firestore/local/leveldb_opener_test.cc

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_opener.h"
+
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/local_serializer.h"
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/autoid.h"
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/test/firebase/firestore/testutil/filesystem_testing.h"
+#include "Firestore/core/test/firebase/firestore/testutil/status_testing.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using core::DatabaseInfo;
+using model::DatabaseId;
+using remote::Serializer;
+using testutil::IsNotFound;
+using testutil::IsOk;
+using testutil::TestTempDir;
+using util::CreateAutoId;
+using util::IsDirectory;
+using util::Path;
+using util::Status;
+using util::TempDir;
+
+DatabaseInfo FakeDatabaseInfo() {
+  return DatabaseInfo(testutil::DbId(), "key", "example.com", true);
+}
+
+DatabaseInfo FakeDatabaseInfoOtherProject() {
+  return DatabaseInfo(testutil::DbId("other-project"), "key", "example.com",
+                      true);
+}
+
+void RunPersistence(LevelDbOpener* opener) {
+  auto created = opener->Create(LruParams::Disabled());
+
+  EXPECT_OK(created.status());
+  auto persistence = std::move(created).ValueOrDie();
+  persistence->Shutdown();
+}
+
+}  // namespace
+
+TEST(LevelDbOpenerTest, CanFindAppDataDir) {
+  LevelDbOpener opener(FakeDatabaseInfo());
+  Path path = opener.AppDataDir();
+  ASSERT_OK(opener.status());
+  EXPECT_THAT(path.Basename().ToUtf8String(), testing::EndsWith("firestore"));
+}
+
+TEST(LevelDbOpenerTest, CanFindLegacyDocumentsDir) {
+  LevelDbOpener opener(FakeDatabaseInfo());
+  Path path = opener.LegacyDocumentsDir();
+  EXPECT_OK(opener.status());
+  EXPECT_THAT(path.Basename().ToUtf8String(), testing::EndsWith("firestore"));
+}
+
+TEST(LevelDbOpenerTest, CanMigrateLegacyData) {
+  TestTempDir root_dir;
+
+  // These names don't actually matter, and work on any platform
+  Path legacy_dir = root_dir.Child("Documents/firestore");
+  Path new_dir = root_dir.Child("Library/Application Support/firestore");
+
+  ASSERT_THAT(IsDirectory(legacy_dir), IsNotFound());
+  ASSERT_THAT(IsDirectory(new_dir), IsNotFound());
+
+  DatabaseInfo db_info = FakeDatabaseInfo();
+  {
+    // Open as if the old way
+    LevelDbOpener opener(db_info);
+    ASSERT_FALSE(opener.PreferredExists(legacy_dir));
```

Comment:
I think this test may be why we have `PreferredExists()` and it has its funky contract?  Perhaps instead we could have  a (test-only) LevelDbOpener constructor that lets you pass in an arbitrary legacy dir and new dir?

Reply:
Passing in an arbitrary legacy new dir turns out to be super annoying because there's three states to represent:

  * pre-specified actual directory
  * pre-specified non-directory 
  * not pre-specified, so go to the Filesystem

Short of making the stored value a `StatusOr<Path>` I couldn't come up with a great way to do that.

In any case I think a better way was to just make the filesystem itself pluggable. I've done that instead.